File: selection/Selection.java

import java.util.Arrays;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class Selection {

    public static void main(String[] args) {
        Random random = new Random();
        int[] lista1 = random.ints(10000, 0, 100000).toArray();

        int[] ordenado = outPlace(lista1);
        System.out.println(Arrays.toString(ordenado));

        int[] lista2 = Arrays.copyOf(lista1, lista1.length);
        inPlace(lista2);
        System.out.println(Arrays.toString(lista2));

        System.out.println(Arrays.equals(ordenado, lista2));
    }

    private static void inPlace(int[] lista) {
        for (int i = 0; i < lista.length; i++) {
                int menorIndice = i;
                for (int j = i + 1; j < lista.length; j++) {
                        if (lista[j] < lista[menorIndice]) {
                            menorIndice = j;
                        }
                }
                int antigo = lista[i];
                lista[i] = lista[menorIndice];
                lista[menorIndice] = antigo;
        }
    }

    private static int[] outPlace(int[] lista) {
        int[] ordenado = new int[lista.length];
        Set<Integer> vistos = new HashSet<>();

        int maiorIndice = maiorIdx(lista);
        for (int i = 0; i < lista.length; i++) {
            int menorIndice = maiorIndice;
            for (int j = 0; j < lista.length; j++) {
                    if (lista[j] <= lista[menorIndice] && !vistos.contains(j)) {
                        menorIndice = j;
                    }
            }
            vistos.add(menorIndice);
            ordenado[i] = lista[menorIndice];
        }

        return ordenado;
    }

    private static int maiorIdx(int[] lista) {
        if (lista.length == 0) return Integer.MIN_VALUE;
        int maior = 0;
        for (int i = 1; i < lista.length; i++) {
                if (lista[i] > lista[maior]) {
                    maior = i;
                }
        }
        return maior;
    }
}

File: selection/selection.py
from random import randint
from itertools import repeat

def in_place(lista):
    for i in range(len(lista)):
        menor_ind = i
        for j in range(i + 1, len(lista)):
            if lista[j] < lista[menor_ind]:
                menor_ind = j
        lista[menor_ind], lista[i] = lista[i], lista[menor_ind]


def out_place(lista):
    ordenado = []
    copia = lista.copy()

    while copia:
        menor_ind = 0
        for i in range(len(copia)):
            if copia[i] < copia[menor_ind]:
                menor_ind = i
        ordenado.append(copia.pop(menor_ind))

    return ordenado


if __name__ == '__main__':
    lista1 = [randint(0, 10000) for _ in repeat(None, 1000)]
    print(out_place(lista1))

    lista2 = lista1.copy()
    in_place(lista2)
    print(lista2)

    lista3 = lista1.copy()
    lista3.sort()
    assert lista2 == lista3